lib/ecto/type.ex

```diff
@@ -291,10 +293,19 @@ defmodule Ecto.Type do
     {:ok, %Ecto.Query.Tagged{value: nil, type: type(type)}}
   end
 
-  def dump({:array, type}, value) do
+  def dump({:array, {:map, _}}, []) do
+    {:ok, %Ecto.Query.Tagged{value: [], type: :map}}
+  end
+
+  def dump({:array, type}, value) when is_list(value) do
     dump_array(type, value, [], false)
   end
 
+  def dump({:map, model}, %{__struct__: model} = struct) do
+    value = Map.take(struct, model.__schema__(:fields))
+    {:ok, %Ecto.Query.Tagged{value: value, type: :map}}
```

Comment:
Why not simply return the map? If we are going to start tagging maps, then we should be consistent and tag them all.